source/simulator/nullspace.cc

```diff
@@ -81,162 +90,213 @@ namespace aspect
-               ExcNotImplemented());
-
-        net_rotations_translations.resize(funcs.size());
-        for (unsigned int i=0; i<funcs.size(); ++i)
-          {
-            // for each of the null space dimensions, set up
-            // a vector, fill it with an element of the null space,
-            // and normalize it
-            net_rotations_translations[i].reinit(
-              introspection.index_sets.system_partitioning[introspection.block_indices.velocities],
-              mpi_communicator);
-            interpolate_onto_velocity_system(*funcs[i],
-                                             net_rotations_translations[i]);
-            net_rotations_translations[i] /= net_rotations_translations[i].l2_norm();
-          }
+    QGauss<dim> quadrature(parameters.stokes_velocity_degree+1);
+    FEValues<dim> fe(mapping, finite_element, quadrature,
+                     UpdateFlags(update_quadrature_points | update_JxW_values | update_values));
 
-      }
-  }
+    typename DoFHandler<dim>::active_cell_iterator cell;
+    std::vector<Vector<double> > fe_vals(quadrature.size(), Vector<double>(finite_element.n_components()));
+
+    Tensor<1,dim> local_momentum;
+    double local_mass = 0.0;
+
+    //loop over all local cells
+    for (cell = dof_handler.begin_active(); cell != dof_handler.end(); ++cell)
+      if (cell->is_locally_owned())
+        {
+          fe.reinit (cell);
+          const std::vector<Point<dim> > &q_points = fe.get_quadrature_points();
+          fe.get_function_values(relevant_dst, fe_vals);
 
+          // get the density at each quadrature point
+          typename MaterialModel::Interface<dim>::MaterialModelInputs in(q_points.size(),
+                                                                parameters.n_compositional_fields);
+          typename MaterialModel::Interface<dim>::MaterialModelOutputs out(q_points.size(),
+                                                                  parameters.n_compositional_fields);
+          if ( ! use_constant_density)
+            {
+              for (unsigned int i=0; i< q_points.size(); i++)
+                {
+                  in.pressure[i] = fe_vals[i][introspection.component_indices.pressure];
+                  in.temperature[i] = fe_vals[i][introspection.component_indices.temperature];
+                  for (unsigned int c=0; c<parameters.n_compositional_fields; ++c)
+                    in.composition[i][c] = fe_vals[i][introspection.component_indices.compositional_fields[c]];
+                  in.position[i] = q_points[i];
+                }
+              material_model->evaluate(in, out);
+            }
 
+          Tensor<1,dim> velocity;
+          // actually compute the momentum and mass
+          for (unsigned int k=0; k< quadrature.size(); ++k)
+            {
+              // get the velocity and density at this quadrature point
+              for (unsigned int i=0; i<dim; ++i) velocity[i] = fe_vals[k][i];
```

Comment:
this assumes velocity starts at component 0. Can you add an assert for this at the beginning of this function?